src/components/tl/cuda/tl_cuda_topo.c

```diff
@@ -5,28 +5,99 @@
+               &pci_id->device, &pci_id->function);
+    if (n != 4) {
+        return UCC_ERR_INVALID_PARAM;
+    }
+    return UCC_OK;
+}
+
+static void ucc_tl_cuda_topo_pci_id_to_str(const ucc_tl_cuda_device_id_t *pci_id,
+                                           char *str, size_t max)
+{
+    ucc_snprintf_safe(str, max, "%04x:%02x:%02x.%d", pci_id->domain,
+                      pci_id->bus, pci_id->device, pci_id->function);
+}
+
+ucc_status_t ucc_tl_cuda_topo_get_pci_id(const ucc_base_lib_t *lib,
+                                         int device,
+                                         ucc_tl_cuda_device_id_t *pci_id)
+{
+    char pci_bus_id[MAX_PCI_BUS_ID_STR];
+    ucc_status_t st;
+
+    CUDACHECK_GOTO(cudaDeviceGetPCIBusId(pci_bus_id, MAX_PCI_BUS_ID_STR,
+                                         device), exit, st, lib);
+    st = ucc_tl_cuda_topo_pci_id_from_str(pci_bus_id, pci_id);
+exit:
+    return st;
+}
 
 static ucc_status_t ucc_tl_cuda_topo_init_matrix(ucc_tl_cuda_team_t *team,
                                                  ucc_tl_cuda_topo_t *cuda_topo)
 {
-    ucc_rank_t size = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t   size   = UCC_TL_TEAM_SIZE(team);
+    ucc_status_t status = UCC_OK;
+    int num_nvlinks, link;
     ucc_rank_t i, j;
-    int dev, peer_dev, peer_access;
-    ucc_status_t status;
-
+    nvmlDevice_t nvml_dev;
+    nvmlFieldValue_t nvml_value;
+    nvmlPciInfo_t nvml_pci;
+    char pci_bus_str[MAX_PCI_BUS_ID_STR];
+
+    nvml_value.fieldId = NVML_FI_DEV_NVLINK_LINK_COUNT;
+    memset(cuda_topo->matrix, 0, sizeof(*cuda_topo->matrix) * size * size);
+    NVMLCHECK_GOTO(nvmlInit_v2(), exit_return, status, UCC_TL_TEAM_LIB(team));
     for (i = 0; i < size; i++) {
-       dev = team->ids[i].device;
-       cuda_topo->matrix[i*size + i] = 1;
-        for (j = i + 1; j < size; j++) {
-            peer_dev = team->ids[j].device;
-            CUDACHECK_GOTO(cudaDeviceCanAccessPeer(&peer_access, dev, peer_dev),
-                           exit_err, status, UCC_TL_TEAM_LIB(team));
-            cuda_topo->matrix[i*size + j] = peer_access;
-            cuda_topo->matrix[j*size + i] = peer_access;
+        cuda_topo->matrix[i*size + i] = 1;
+        ucc_tl_cuda_topo_pci_id_to_str(&team->ids[i].pci_id, pci_bus_str,
+                                       MAX_PCI_BUS_ID_STR);
+        NVMLCHECK_GOTO(nvmlDeviceGetHandleByPciBusId_v2(pci_bus_str, &nvml_dev),
+                       exit_nvml_shutdown, status, UCC_TL_TEAM_LIB(team));
+        NVMLCHECK_GOTO(nvmlDeviceGetFieldValues(nvml_dev, 1, &nvml_value),
+                       exit_nvml_shutdown, status, UCC_TL_TEAM_LIB(team));
+        num_nvlinks = ((nvml_value.nvmlReturn == NVML_SUCCESS) &&
+                       (nvml_value.valueType == NVML_VALUE_TYPE_UNSIGNED_INT)) ?
+                      nvml_value.value.uiVal : 0;
+        for (link = 0 ; link < num_nvlinks; link++) {
+            NVMLCHECK_GOTO(nvmlDeviceGetNvLinkRemotePciInfo_v2(nvml_dev, link,
```

Comment:
is this a valid  in case of NVSwitch?

Reply:
@bureddy right, it didn't work for NVSwitch. I had to modify tl cuda topology structs in order to support it (updated PR).
Now we have tl cuda context topology that contains graph of all cuda devices on the node including GPUs and NVSwtiches. Using this graph team topology is computed during team create.
I didn't add NVSwitch case to ucc_tl_cuda_topo_num_links() function to make this PR smaller so NVSwtich is not supported yet, but it won't be hard to add it now.